metadata.sh

```diff
@@ -0,0 +1,140 @@
+#!/usr/bin/env bash
+set -Eeuo pipefail
+
+workdir="$(readlink -f "$BASH_SOURCE")"
+workdir="$(dirname "$workdir")"
+#cd "$workdir"
+
+jsonFile='metadata.json'
+canonicalMetadataFile="$workdir/$jsonFile"
+maxCategories=3
+
+self="$(basename "$0")"
+
+usage() {
+	cat <<EOUSAGE
+
+usage: $self [--categories] [--diff] [--fmt] repo [...]
+   ie: $self debian
+       $self -d */
+       $self -dc python
+
+This script checks a repo's metadata.json and checks categories, diffs, or formats it.
+
+-c, --categories      Check that the categories are from the valid set of categories and that there are no more than $maxCategories. Exits non-zero if there are too many categories or they are unkown.
+-d, --diff            Check formatting of the '[repo]/metadata.json' and print a diff. Default action if no flags are supplied. Exits non-zero if there is a difference.
+-f, --fmt, --format   Apply the formatting that the '-d' flag would output.
+-h, --help            Print this help output and exit.
+
+Arguments are the list of repos with a 'metadata.json' in them. 'metadata.json' is expected in every repo
+'.' can also be passed to check the format of the canonical './metadata.json' at
+the root of the repo, but the max categories of '-c' is skipped for it.
+EOUSAGE
+}
+
+# arg handling
+opts="$(getopt -o 'cdfh' --long 'categories,diff,fmt,format,help' -- "$@" || { usage >&2 && false; })"
+eval set -- "$opts"
+
+categories=
+diff=
+fmt=
+
+while true; do
+	flag="$1"
+	shift
+	case "$flag" in
+		--categories|-c) categories=1 ;;
+		--diff|-d) diff=1 ;;
+		--fmt|--format|-f) fmt=1 ;;
+		--help|-h) usage && exit 0 ;;
+		--) break ;;
+		*)
+			{
+				echo "error: unknown flag: $flag"
+				usage
+			} >&2
+			exit 1
+			;;
+	esac
+done
+
+# default to print diff
+if [ -z "$diff" ] && [ -z "$categories" ] && [ -z "$fmt" ]; then
+	diff=1
+fi
+
+repos=( "$@" )
+if [ "${#repos[@]}" -eq 0 ]; then
+	repos=( */ )
+fi
+repos=( "${repos[@]%/}" )
+
+canonicalMetadataFileJson="$(cat "$canonicalMetadataFile")"
+export canonicalMetadataFileJson
```

Comment:
I thought slurpfile would be slower, since we'd have to open/read the canonical file for every repo, but seems negligible locally, so I'm fine with it.